File: backend/internal/handlers/handlers.go
package handlers

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/continuum/backend/internal/grpc"
	"github.com/gorilla/mux"
)

// ValidationError represents a structured validation error
type ValidationError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
	Code    string `json:"code"`
}

// ErrorResponse represents a structured error response
type ErrorResponse struct {
	Error      string            `json:"error"`
	Message    string            `json:"message,omitempty"`
	Errors     []ValidationError `json:"errors,omitempty"`
	Timestamp  int64             `json:"timestamp"`
	RequestID  string            `json:"request_id,omitempty"`
}

// RequestLimits defines various request size and validation limits
type RequestLimits struct {
	MaxRequestSize    int64         // Maximum request body size in bytes
	MaxResponseSize   int64         // Maximum response size in bytes
	Timeout           time.Duration // Request timeout
	MaxTickNumber     uint64        // Maximum valid tick number
	MaxRecentTicks    int           // Maximum number of recent ticks to return
}

// Handler struct with optimized HTTP client and validation
type Handler struct {
	grpcClient   *grpc.Client
	restBaseURL  string
	httpClient   *http.Client // Shared HTTP client with connection pooling
	limits       RequestLimits
	txHashRegex  *regexp.Regexp // Compiled regex for transaction hash validation
}

// Default limits for security and performance
var defaultLimits = RequestLimits{
	MaxRequestSize:  1024 * 1024,    // 1MB
	MaxResponseSize: 10 * 1024 * 1024, // 10MB
	Timeout:         30 * time.Second,
	MaxTickNumber:   1000000000, // 1 billion
	MaxRecentTicks:  1000,
}

// NewHandler creates a new handler with optimized HTTP client and validation
func NewHandler(grpcClient *grpc.Client, restBaseURL string) *Handler {
	// Create optimized HTTP client with connection pooling
	transport := &http.Transport{
		// Connection pooling configuration
		MaxIdleConns:        100,              // Maximum idle connections across all hosts
		MaxIdleConnsPerHost: 10,               // Maximum idle connections per host
		MaxConnsPerHost:     50,               // Maximum connections per host
		IdleConnTimeout:     90 * time.Second, // How long an idle connection is kept
		TLSHandshakeTimeout: 10 * time.Second, // TLS handshake timeout
		ExpectContinueTimeout: 1 * time.Second, // Expect: 100-continue timeout
		
		// TCP connection settings
		DialContext: (&net.Dialer{
			Timeout:   10 * time.Second, // Connection timeout
			KeepAlive: 30 * time.Second, // TCP keep-alive
		}).DialContext,
		
		// Security settings
		TLSClientConfig: &tls.Config{
			MinVersion: tls.VersionTLS12, // Minimum TLS 1.2
		},
		
		// Response header timeout
		ResponseHeaderTimeout: 10 * time.Second,
	}

	httpClient := &http.Client{
		Transport: transport,
		Timeout:   defaultLimits.Timeout,
		// Don't follow redirects for security
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	// Compile transaction hash regex (hexadecimal, 8 characters)
	txHashRegex := regexp.MustCompile(`^[a-fA-F0-9]{8}$`)

	return &Handler{
		grpcClient:  grpcClient,
		restBaseURL: restBaseURL,
		httpClient:  httpClient,
		limits:      defaultLimits,
		txHashRegex: txHashRegex,
	}
}

// validateTransactionHash validates a transaction hash
func (h *Handler) validateTransactionHash(hash string) *ValidationError {
	if hash == "" {
		return &ValidationError{
			Field:   "hash",
			Message: "Transaction hash is required",
			Code:    "required",
		}
	}

	if !h.txHashRegex.MatchString(hash) {
		return &ValidationError{
			Field:   "hash",
			Message: "Transaction hash must be exactly 8 hexadecimal characters",
			Code:    "invalid_format",
		}
	}

	return nil
}

// validateTickNumber validates a tick number
func (h *Handler) validateTickNumber(tickStr string) (uint64, *ValidationError) {
	if tickStr == "" {
		return 0, &ValidationError{
			Field:   "number",
			Message: "Tick number is required",
			Code:    "required",
		}
	}

	tickNum, err := strconv.ParseUint(tickStr, 10, 64)
	if err != nil {
		return 0, &ValidationError{
			Field:   "number",
			Message: "Tick number must be a valid positive integer",
			Code:    "invalid_format",
		}
	}

	if tickNum > h.limits.MaxTickNumber {
		return 0, &ValidationError{
			Field:   "number",
			Message: fmt.Sprintf("Tick number must not exceed %d", h.limits.MaxTickNumber),
			Code:    "out_of_range",
		}
	}

	return tickNum, nil
}

// validateQueryParams validates query parameters for recent ticks endpoint
func (h *Handler) validateQueryParams(r *http.Request) []ValidationError {
	var errors []ValidationError

	// Validate limit parameter
	if limitStr := r.URL.Query().Get("limit"); limitStr != "" {
		limit, err := strconv.Atoi(limitStr)
		if err != nil {
			errors = append(errors, ValidationError{
				Field:   "limit",
				Message: "Limit must be a valid integer",
				Code:    "invalid_format",
			})
		} else if limit < 1 {
			errors = append(errors, ValidationError{
				Field:   "limit",
				Message: "Limit must be greater than 0",
				Code:    "out_of_range",
			})
		} else if limit > h.limits.MaxRecentTicks {
			errors = append(errors, ValidationError{
				Field:   "limit",
				Message: fmt.Sprintf("Limit must not exceed %d", h.limits.MaxRecentTicks),
				Code:    "out_of_range",
			})
		}
	}

	// Validate offset parameter
	if offsetStr := r.URL.Query().Get("offset"); offsetStr != "" {
		offset, err := strconv.ParseUint(offsetStr, 10, 64)
		if err != nil {
			errors = append(errors, ValidationError{
				Field:   "offset",
				Message: "Offset must be a valid non-negative integer",
				Code:    "invalid_format",
			})
		} else if offset > h.limits.MaxTickNumber {
			errors = append(errors, ValidationError{
				Field:   "offset",
				Message: fmt.Sprintf("Offset must not exceed %d", h.limits.MaxTickNumber),
				Code:    "out_of_range",
			})
		}
	}

	return errors
}

// sanitizeInput performs basic input sanitization
func sanitizeInput(input string) string {
	// Remove control characters and normalize whitespace
	input = strings.TrimSpace(input)
	// Remove null bytes and other control characters
	input = strings.ReplaceAll(input, "\x00", "")
	return input
}

// sendErrorResponse sends a structured error response with logging
func (h *Handler) sendErrorResponse(w http.ResponseWriter, r *http.Request, statusCode int, message string, validationErrors []ValidationError) {
	response := ErrorResponse{
		Error:     http.StatusText(statusCode),
		Message:   message,
		Errors:    validationErrors,
		Timestamp: time.Now().Unix(),
	}

	// Log the error with context
	log.Printf("❌ [%s %s] %d: %s", r.Method, r.URL.Path, statusCode, message)
	if len(validationErrors) > 0 {
		log.Printf("   Validation errors: %+v", validationErrors)
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	json.NewEncoder(w).Encode(response)
}

// requestSizeLimitMiddleware enforces request size limits
func (h *Handler) requestSizeLimitMiddleware(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.ContentLength > h.limits.MaxRequestSize {
			h.sendErrorResponse(w, r, http.StatusRequestEntityTooLarge, 
				fmt.Sprintf("Request body too large. Maximum size: %d bytes", h.limits.MaxRequestSize), nil)
			return
		}

		// Limit the request body reader
		r.Body = http.MaxBytesReader(w, r.Body, h.limits.MaxRequestSize)
		next(w, r)
	}
}

// makeSecureRequest makes an HTTP request with security checks and proper error handling
func (h *Handler) makeSecureRequest(ctx context.Context, method, url string, body io.Reader) (*http.Response, error) {
	req, err := http.NewRequestWithContext(ctx, method, url, body)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set security headers
	req.Header.Set("User-Agent", "fermi-explorer-proxy/1.0")
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	// Make the request
	resp, err := h.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}

	// Check response size
	if resp.ContentLength > h.limits.MaxResponseSize {
		resp.Body.Close()
		return nil, fmt.Errorf("response too large: %d bytes", resp.ContentLength)
	}

	return resp, nil
}

// Health check endpoint with enhanced security
func (h *Handler) Health(w http.ResponseWriter, r *http.Request) {
	// Basic method validation
	if r.Method != http.MethodGet {
		h.sendErrorResponse(w, r, http.StatusMethodNotAllowed, "Method not allowed", nil)
		return
	}

	response := map[string]interface{}{
		"status":    "healthy",
		"timestamp": time.Now().Unix(),
		"version":   "1.0.0",
	}
	
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
	json.NewEncoder(w).Encode(response)
}

// Status endpoint - proxies to REST API with enhanced validation and security
func (h *Handler) Status(w http.ResponseWriter, r *http.Request) {
	// Method validation
	if r.Method != http.MethodGet {
		h.sendErrorResponse(w, r, http.StatusMethodNotAllowed, "Method not allowed", nil)
		return
	}

	// Create request context with timeout
	ctx, cancel := context.WithTimeout(r.Context(), h.limits.Timeout)
	defer cancel()

	// Make secure request
	resp, err := h.makeSecureRequest(ctx, http.MethodGet, h.restBaseURL+"/status", nil)
	if err != nil {
		errorMsg := "Failed to get status from sequencer"
		log.Printf("❌ Status endpoint error: %v", err)
		h.sendErrorResponse(w, r, http.StatusServiceUnavailable, errorMsg, nil)
		return
	}
	defer resp.Body.Close()

	// Limit response body size
	limitedReader := io.LimitReader(resp.Body, h.limits.MaxResponseSize)
	
	// Copy the response from the sequencer
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
	w.WriteHeader(resp.StatusCode)
	
	var response interface{}
	if err := json.NewDecoder(limitedReader).Decode(&response); err != nil {
		h.sendErrorResponse(w, r, http.StatusInternalServerError, "Failed to decode sequencer response", nil)
		return
	}
	
	json.NewEncoder(w).Encode(response)
}

// GetTransaction - transaction lookup endpoint with comprehensive validation
func (h *Handler) GetTransaction(w http.ResponseWriter, r *http.Request) {
	// Method validation
	if r.Method != http.MethodGet {
		h.sendErrorResponse(w, r, http.StatusMethodNotAllowed, "Method not allowed", nil)
		return
	}

	vars := mux.Vars(r)
	txHash := sanitizeInput(vars["hash"])

	// Validate transaction hash
	if validationErr := h.validateTransactionHash(txHash); validationErr != nil {
		h.sendErrorResponse(w, r, http.StatusBadRequest, "Invalid transaction hash", []ValidationError{*validationErr})
		return
	}

	// Create request context with timeout
	ctx, cancel := context.WithTimeout(r.Context(), h.limits.Timeout)
	defer cancel()

	// Make secure request
	resp, err := h.makeSecureRequest(ctx, http.MethodGet, h.restBaseURL+"/tx/"+txHash, nil)
	if err != nil {
		log.Printf("❌ Failed to get transaction %s: %v", txHash, err)
		h.sendErrorResponse(w, r, http.StatusInternalServerError, "Failed to get transaction", nil)
		return
	}
	defer resp.Body.Close()

	// Limit response body size
	limitedReader := io.LimitReader(resp.Body, h.limits.MaxResponseSize)

	// Copy the response from the sequencer
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
	w.WriteHeader(resp.StatusCode)
	
	var response interface{}
	if err := json.NewDecoder(limitedReader).Decode(&response); err != nil {
		h.sendErrorResponse(w, r, http.StatusInternalServerError, "Failed to decode sequencer response", nil)
		return
	}
	
	log.Printf("✅ Successfully retrieved transaction: %s", txHash)
	json.NewEncoder(w).Encode(response)
}

// GetTick - tick lookup endpoint with comprehensive validation
func (h *Handler) GetTick(w http.ResponseWriter, r *http.Request) {
	// Method validation
	if r.Method != http.MethodGet {
		h.sendErrorResponse(w, r, http.StatusMethodNotAllowed, "Method not allowed", nil)
		return
	}

	vars := mux.Vars(r)
	tickNumStr := sanitizeInput(vars["number"])

	// Validate tick number
	tickNum, validationErr := h.validateTickNumber(tickNumStr)
	if validationErr != nil {
		h.sendErrorResponse(w, r, http.StatusBadRequest, "Invalid tick number", []ValidationError{*validationErr})
		return
	}

	// Create request context with timeout
	ctx, cancel := context.WithTimeout(r.Context(), h.limits.Timeout)
	defer cancel()

	// Make secure request
	resp, err := h.makeSecureRequest(ctx, http.MethodGet, h.restBaseURL+"/tick/"+tickNumStr, nil)
	if err != nil {
		log.Printf("❌ Failed to get tick %d: %v", tickNum, err)
		h.sendErrorResponse(w, r, http.StatusInternalServerError, "Failed to get tick", nil)
		return
	}
	defer resp.Body.Close()

	// Limit response body size
	limitedReader := io.LimitReader(resp.Body, h.limits.MaxResponseSize)

	// Copy the response from the sequencer
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
	w.WriteHeader(resp.StatusCode)
	
	var response interface{}
	if err := json.NewDecoder(limitedReader).Decode(&response); err != nil {
		h.sendErrorResponse(w, r, http.StatusInternalServerError, "Failed to decode sequencer response", nil)
		return
	}
	
	log.Printf("✅ Successfully retrieved tick: %d", tickNum)
	json.NewEncoder(w).Encode(response)
}

// GetRecentTicks - recent ticks endpoint with comprehensive validation
func (h *Handler) GetRecentTicks(w http.ResponseWriter, r *http.Request) {
	// Method validation
	if r.Method != http.MethodGet {
		h.sendErrorResponse(w, r, http.StatusMethodNotAllowed, "Method not allowed", nil)
		return
	}

	// Validate query parameters
	validationErrors := h.validateQueryParams(r)
	if len(validationErrors) > 0 {
		h.sendErrorResponse(w, r, http.StatusBadRequest, "Invalid query parameters", validationErrors)
		return
	}

	// Create request context with timeout
	ctx, cancel := context.WithTimeout(r.Context(), h.limits.Timeout)
	defer cancel()

	// Forward query parameters safely
	url := h.restBaseURL + "/ticks/recent"
	if r.URL.RawQuery != "" {
		// Sanitize and validate query string
		safeQuery := sanitizeInput(r.URL.RawQuery)
		url += "?" + safeQuery
	}

	// Make secure request
	resp, err := h.makeSecureRequest(ctx, http.MethodGet, url, nil)
	if err != nil {
		log.Printf("❌ Failed to get recent ticks: %v", err)
		h.sendErrorResponse(w, r, http.StatusInternalServerError, "Failed to get recent ticks", nil)
		return
	}
	defer resp.Body.Close()

	// Limit response body size
	limitedReader := io.LimitReader(resp.Body, h.limits.MaxResponseSize)

	// Copy the response from the sequencer
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
	w.WriteHeader(resp.StatusCode)
	
	var response interface{}
	if err := json.NewDecoder(limitedReader).Decode(&response); err != nil {
		h.sendErrorResponse(w, r, http.StatusInternalServerError, "Failed to decode sequencer response", nil)
		return
	}
	
	log.Printf("✅ Successfully retrieved recent ticks")
	json.NewEncoder(w).Encode(response)
}


// GetRequestSizeLimitMiddleware returns the request size limit middleware
// This can be used in the main.go to wrap endpoints that need request size validation
func (h *Handler) GetRequestSizeLimitMiddleware() func(http.HandlerFunc) http.HandlerFunc {
	return h.requestSizeLimitMiddleware
}

File: apps/backend/src/handlers/handlers.ts
import { Context } from "hono";
import {
  GrpcClient,
  GetStatusResponse,
  GetTransactionResponse,
  GetTickResponse,
  GetChainStateResponse,
  SubmitBatchResponse,
} from "../grpc/client";
import {
  validateTransactionHash,
  validateTickNumber,
  validateQueryParams,
  sanitizeInput,
  sendErrorResponse,
  limits,
} from "../middleware/validation";
import { MemoryCache } from "../cache/memory-cache";

export class Handler {
  private grpcClient: GrpcClient;
  private restBaseURL: string;
  private matchEngineURL: string;
  private tickCache: MemoryCache<any>;
  private transactionCache: MemoryCache<any>;

  constructor(
    grpcClient: GrpcClient,
    restBaseURL: string,
    matchEngineURL: string
  ) {
    this.grpcClient = grpcClient;
    this.restBaseURL = restBaseURL;
    this.matchEngineURL = matchEngineURL;
    this.tickCache = new MemoryCache<any>(10); // 10 minutes TTL
    this.transactionCache = new MemoryCache<any>(30); // 30 minutes TTL
    
    // Setup periodic cache cleanup
    this.setupCacheCleanup();
  }

  private setupCacheCleanup(): void {
    // Clean up expired cache entries every 5 minutes
    setInterval(() => {
      const tickCleaned = this.tickCache.cleanup();
      const txCleaned = this.transactionCache.cleanup();
      if (tickCleaned + txCleaned > 0) {
        console.log(`🧹 Cache cleanup: removed ${tickCleaned} ticks and ${txCleaned} transactions`);
      }
    }, 5 * 60 * 1000); // 5 minutes
  }

  private async makeSecureRequest(
    method: string,
    url: string,
    body?: any
  ): Promise<Response> {
    const headers: Record<string, string> = {
      "User-Agent": "fermi-explorer-proxy/1.0",
      Accept: "application/json",
    };

    if (body) {
      headers["Content-Type"] = "application/json";
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), limits.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers,
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      return response;
    } catch (error) {
      clearTimeout(timeoutId);
      throw error;
    }
  }

  async health(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    const response = {
      status: "healthy",
      timestamp: Math.floor(Date.now() / 1000),
      version: "1.0.0",
    };

    return c.json(response, {
      headers: {
        "Cache-Control": "no-cache, no-store, must-revalidate",
      },
    });
  }

  async status(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    try {
      const statusResponse: GetStatusResponse =
        await this.grpcClient.getStatus();

      return c.json(statusResponse, {
        headers: {
          "Cache-Control": "no-cache, no-store, must-revalidate",
        },
      });
    } catch (error) {
      console.error("❌ Status endpoint error:", error);
      return sendErrorResponse(c, 503, "Failed to get status from sequencer");
    }
  }

  async getTransaction(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    const txHash = sanitizeInput(c.req.param("hash") || "");

    const validationError = validateTransactionHash(txHash);
    if (validationError) {
      return sendErrorResponse(c, 400, "Invalid transaction hash", [
        validationError,
      ]);
    }

    // Check cache first
    const cacheKey = `tx:${txHash}`;
    const cachedData = this.transactionCache.get(cacheKey);
    if (cachedData) {
      console.log(`✅ Cache hit for transaction: ${txHash}`);
      return c.json(cachedData, {
        headers: {
          "Cache-Control": "private, max-age=1800", // 30 minutes
          "X-Cache": "HIT",
        },
      });
    }

    try {
      const resp = await this.makeSecureRequest(
        "GET",
        `${this.restBaseURL}/tx/${txHash}`
      );

      if (!resp.ok) {
        if (resp.status === 404) {
          return sendErrorResponse(c, 404, "Transaction not found");
        }
        return sendErrorResponse(c, 502, "Failed to get transaction");
      }

      const data = await resp.json();
      console.log(`✅ Successfully retrieved transaction: ${txHash}`);

      // Cache the response
      this.transactionCache.set(cacheKey, data);

      return c.json(data, {
        headers: {
          "Cache-Control": "private, max-age=1800", // 30 minutes
          "X-Cache": "MISS",
        },
      });
    } catch (error) {
      console.error(`❌ Failed to get transaction ${txHash}:`, error);
      return sendErrorResponse(c, 500, "Failed to get transaction");
    }
  }

  async getTick(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    const tickNumStr = sanitizeInput(c.req.param("number") || "");

    const { value: tickNum, error: validationError } =
      validateTickNumber(tickNumStr);
    if (validationError) {
      return sendErrorResponse(c, 400, "Invalid tick number", [
        validationError,
      ]);
    }

    // Check cache first
    const cacheKey = `tick:${tickNumStr}`;
    const cachedData = this.tickCache.get(cacheKey);
    if (cachedData) {
      console.log(`✅ Cache hit for tick: ${tickNum}`);
      return c.json(cachedData, {
        headers: {
          "Cache-Control": "private, max-age=600", // 10 minutes
          "X-Cache": "HIT",
        },
      });
    }

    try {
      const resp = await this.makeSecureRequest(
        "GET",
        `${this.restBaseURL}/tick/${tickNumStr}`
      );

      if (!resp.ok) {
        if (resp.status === 404) {
          return sendErrorResponse(c, 404, "Tick not found");
        }
        return sendErrorResponse(c, 502, "Failed to get tick");
      }

      const data = await resp.json();
      console.log(`✅ Successfully retrieved tick: ${tickNum}`);

      // Cache the response
      this.tickCache.set(cacheKey, data);

      return c.json(data, {
        headers: {
          "Cache-Control": "private, max-age=600", // 10 minutes
          "X-Cache": "MISS",
        },
      });
    } catch (error) {
      console.error(`❌ Failed to get tick ${tickNum}:`, error);
      return sendErrorResponse(c, 500, "Failed to get tick");
    }
  }

  async getRecentTicks(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    const validationErrors = validateQueryParams(c);
    if (validationErrors.length > 0) {
      return sendErrorResponse(
        c,
        400,
        "Invalid query parameters",
        validationErrors
      );
    }

    try {
      const url = new URL(c.req.url);
      let targetUrl = `${this.restBaseURL}/ticks/recent`;

      if (url.search) {
        const sanitizedQuery = sanitizeInput(url.search.substring(1));
        targetUrl += `?${sanitizedQuery}`;
      }

      const resp = await this.makeSecureRequest("GET", targetUrl);

      if (!resp.ok) {
        return sendErrorResponse(c, 502, "Failed to get recent ticks");
      }

      const data = await resp.json();
      console.log("✅ Successfully retrieved recent ticks");

      return c.json(data, {
        headers: {
          "Cache-Control": "no-cache, no-store, must-revalidate",
        },
      });
    } catch (error) {
      console.error("❌ Failed to get recent ticks:", error);
      return sendErrorResponse(c, 500, "Failed to get recent ticks");
    }
  }

  async submitTransaction(c: Context): Promise<Response> {
    if (c.req.method !== "POST") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    try {
      const body = await c.req.json();
      console.log("bodyReceived", body);
      // Validate transaction structure according to spec
      if (!body.transaction) {
        return sendErrorResponse(c, 400, "Transaction object is required");
      }

      const transaction = body.transaction;

      // Validate required fields
      if (!transaction.tx_id || typeof transaction.tx_id !== "string") {
        return sendErrorResponse(
          c,
          400,
          "tx_id is required and must be a string"
        );
      }

      if (!transaction.signature) {
        return sendErrorResponse(
          c,
          400,
          "signature is required and must be a byte array"
        );
      }

      if (!transaction.public_key) {
        return sendErrorResponse(
          c,
          400,
          "public_key is required and must be a byte array"
        );
      }

      if (
        transaction.nonce === undefined ||
        typeof transaction.nonce !== "number"
      ) {
        return sendErrorResponse(
          c,
          400,
          "nonce is required and must be a number"
        );
      }

      if (!transaction.timestamp) {
        return sendErrorResponse(
          c,
          400,
          "timestamp is required and must be a number"
        );
      }

      // Prepare transaction for gRPC client
      const grpcTransaction = {
        tx_id: transaction.tx_id,
        payload: transaction.payload,
        signature: transaction.signature,
        public_key: transaction.public_key,
        nonce: transaction.nonce.toString(),
        timestamp: transaction.timestamp.toString(),
      };

      console.log("grpcPayload", grpcTransaction);

      // Submit transaction using gRPC client
      const response = await this.grpcClient.submitTransaction(grpcTransaction);

      console.log("✅ Successfully submitted transaction", response);

      // Metrics removed

      return c.json(response);
    } catch (error) {
      console.error("❌ Failed to submit transaction:", error);

      if (error instanceof SyntaxError) {
        return sendErrorResponse(c, 400, "Invalid JSON in request body");
      }
      return sendErrorResponse(c, 500, "Failed to submit transaction");
    }
  }

  async submitBatch(c: Context): Promise<Response> {
    if (c.req.method !== "POST") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    try {
      const body = await c.req.json();
      console.log("submitBatch:body", body);

      // Basic validation of batch structure
      if (!body.transactions || !Array.isArray(body.transactions)) {
        return sendErrorResponse(c, 400, "Transactions array is required");
      }

      if (body.transactions.length === 0) {
        return sendErrorResponse(c, 400, "Transactions array cannot be empty");
      }

      // Validate and prepare each transaction
      const grpcTransactions = [];
      for (let i = 0; i < body.transactions.length; i++) {
        const transaction = body.transactions[i];

        // Validate required fields for each transaction
        if (!transaction.tx_id || typeof transaction.tx_id !== "string") {
          return sendErrorResponse(
            c,
            400,
            `Transaction ${i}: tx_id is required and must be a string`
          );
        }

        if (!transaction.payload || !Array.isArray(transaction.payload)) {
          return sendErrorResponse(
            c,
            400,
            `Transaction ${i}: payload is required and must be a byte array`
          );
        }

        if (!transaction.signature || !Array.isArray(transaction.signature)) {
          return sendErrorResponse(
            c,
            400,
            `Transaction ${i}: signature is required and must be a byte array`
          );
        }

        if (!transaction.public_key || !Array.isArray(transaction.public_key)) {
          return sendErrorResponse(
            c,
            400,
            `Transaction ${i}: public_key is required and must be a byte array`
          );
        }

        if (
          transaction.nonce === undefined ||
          typeof transaction.nonce !== "number"
        ) {
          return sendErrorResponse(
            c,
            400,
            `Transaction ${i}: nonce is required and must be a number`
          );
        }

        if (
          !transaction.timestamp ||
          typeof transaction.timestamp !== "number"
        ) {
          return sendErrorResponse(
            c,
            400,
            `Transaction ${i}: timestamp is required and must be a number`
          );
        }

        grpcTransactions.push({
          tx_id: transaction.tx_id,
          payload: new Uint8Array(transaction.payload),
          signature: new Uint8Array(transaction.signature),
          public_key: new Uint8Array(transaction.public_key),
          nonce: transaction.nonce.toString(),
          timestamp: transaction.timestamp.toString(),
        });
      }

      // Submit batch using gRPC client
      const response = await this.grpcClient.submitBatch(grpcTransactions);

      console.log(
        `✅ Successfully submitted ${body.transactions.length} transactions`
      );

      return c.json(response);
    } catch (error) {
      console.error("❌ Failed to submit batch transactions:", error);

      if (error instanceof SyntaxError) {
        return sendErrorResponse(c, 400, "Invalid JSON in request body");
      }
      return sendErrorResponse(c, 500, "Failed to submit batch transactions");
    }
  }

  async getChainState(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    try {
      // Get optional tick_limit parameter
      const tickLimitStr = c.req.query("tick_limit");
      let tickLimit: number | undefined;

      if (tickLimitStr) {
        const parsed = parseInt(tickLimitStr, 10);
        if (isNaN(parsed) || parsed < 0) {
          return sendErrorResponse(c, 400, "Invalid tick_limit parameter");
        }
        tickLimit = parsed;
      }

      const chainState: GetChainStateResponse =
        await this.grpcClient.getChainState(tickLimit);

      return c.json(chainState, {
        headers: {
          "Cache-Control": "no-cache, no-store, must-revalidate",
        },
      });
    } catch (error) {
      console.error("❌ Chain state endpoint error:", error);
      return sendErrorResponse(
        c,
        503,
        "Failed to get chain state from sequencer"
      );
    }
  }

  async getMarkets(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    try {
      const resp = await this.makeSecureRequest(
        "GET",
        `${this.matchEngineURL}/markets`
      );

      if (!resp.ok) {
        return sendErrorResponse(c, 502, "Failed to get markets");
      }

      const data = await resp.json();
      console.log("✅ Successfully retrieved markets");

      return c.json(data, {
        headers: {
          "Cache-Control": "no-cache, no-store, must-revalidate",
        },
      });
    } catch (error) {
      console.error("❌ Failed to get markets:", error);
      return sendErrorResponse(c, 500, "Failed to get markets");
    }
  }

  async getMarketOrderbook(c: Context): Promise<Response> {
    if (c.req.method !== "GET") {
      return sendErrorResponse(c, 405, "Method not allowed");
    }

    const marketId = sanitizeInput(c.req.param("marketId") || "");

    if (!marketId) {
      return sendErrorResponse(c, 400, "Market ID is required");
    }

    try {
      const resp = await this.makeSecureRequest(
        "GET",
        `${this.matchEngineURL}/markets/${marketId}/orderbook`
      );

      if (!resp.ok) {
        if (resp.status === 404) {
          return sendErrorResponse(c, 404, "Market not found");
        }
        return sendErrorResponse(c, 502, "Failed to get market orderbook");
      }

      const data = await resp.json();
      console.log(
        `✅ Successfully retrieved orderbook for market: ${marketId}`
      );

      // Metrics removed

      return c.json(data, {
        headers: {
          "Cache-Control": "no-cache, no-store, must-revalidate",
        },
      });
    } catch (error) {
      console.error(
        `❌ Failed to get orderbook for market ${marketId}:`,
        error
      );
      return sendErrorResponse(c, 500, "Failed to get market orderbook");
    }
  }
}
